### PR TITLE
Fix ros2 ball filter

### PR DESCRIPTION
**Pull request by val-ba**
## Proposed changes
- reintroduce necessary outcommented line
- add test script

## Necessary checks
- [ ] Update package version
- [ ] Run `catkin build`
- [ ] Write documentation
- [ ] Create issues for future work
- [ ] Test on your machine
- [ ] Test on the robot
- [ ] Put the PR on our Project board

### CODE REVIEW

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 1]
```diff
@@ -28,11 +28,11 @@ def __init__(self):
         super().__init__("ball_filter", automatically_declare_parameters_from_overrides=True)
         self.tf_buffer = tf2.Buffer(cache_time=rclpy.duration.Duration(seconds=2))
         self.tf_listener = tf2.TransformListener(self.tf_buffer, self)
-
+        self.get_logger().warning("init")
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py]: Remove debug warning

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 2]
```diff
@@ -41,6 +41,7 @@ def _dynamic_reconfigure_callback(self, config):
         :param config: New _config
         :param level: The level is a definable int in the Vision.cfg file. All changed params are or ed together by dynamic reconfigure.
         """
+        self.get_logger().warning("dynamic reconfigure")
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py]: Same here

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 3]
```diff
@@ -110,8 +111,9 @@ def _dynamic_reconfigure_callback(self, config):
         self.filter_timer = self.create_timer(self.filter_time_step, self.filter_step)
         return SetParametersResult(successful=True)
 
-    def ball_callback(self, msg: PoseWithCertaintyArray):
+    def ball_callback(self, msg: PoseWithCertaintyArray, optional_distance=True):
         """handles incoming ball messages"""
+        self.get_logger().warning("Ball callback")
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py]: Same here

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 4]
```diff
@@ -110,8 +111,9 @@ def _dynamic_reconfigure_callback(self, config):
         self.filter_timer = self.create_timer(self.filter_time_step, self.filter_step)
         return SetParametersResult(successful=True)
 
-    def ball_callback(self, msg: PoseWithCertaintyArray):
+    def ball_callback(self, msg: PoseWithCertaintyArray, optional_distance=True):
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py:112]: what does `optional_distance=True` do?

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 5]
```diff
@@ -128,7 +130,9 @@ def ball_callback(self, msg: PoseWithCertaintyArray):
             except (tf2.ConnectivityException, tf2.LookupException, tf2.ExtrapolationException) as e:
                 self.get_logger().warning(str(e))
 
+
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py]: Only one line between functions in a class

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 6]
```diff
@@ -128,7 +130,9 @@ def ball_callback(self, msg: PoseWithCertaintyArray):
...
     def reset_filter_cb(self, req, response):
+        self.get_logger().warning("reset filter calback")
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py]: Same here

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 7]
```diff
@@ -193,6 +197,7 @@ def init_filter(self, x, y):
         :param x: start x position of the ball
         :param y: start y position of the ball
         """
+        self.get_logger().warning("init filter")
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py]: Same here

File: bitbots_ball_filter/bitbots_ball_filter/ball_filter.py [hunk 8]
```diff
@@ -216,6 +221,7 @@ def init_filter(self, x, y):
 
         self.filter_initialized = True
 
+
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_filter.py]: Same here

File: bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+        # setup publisher to publish same ball
+        self.ball_publisher = self.create_publisher(
+            PoseWithCertaintyArray,
+            'balls_relative',
+            1
+        )
+        rate = self.create_timer(0.1, self.publish_ball)
+
+        self.counter = 0
+
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py:21]: One line

File: bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+        array.header.stamp = self.get_clock().now().to_msg()
+        array.header.frame_id = "odom"
+        pose = PoseWithCertainty()
+        pose.pose.pose.position.x = 1.0 + math.sin(self.counter)
+        pose.pose.pose.position.y = 1.0 + math.cos(self.counter)
+        self.counter += 0.1
+        pose.confidence = 1.0
+        array.poses = [pose]
+        self.ball_publisher.publish(array)
+
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py:33]: Two lines

File: bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
+import math
```
Comment (Flova) [inline bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py:1]: Maybe merge `ball_publisher.py` and `ball_sim.py`

File: bitbots_ball_filter/scripts/ball_sim.py [hunk 12]
```diff
@@ -8,7 +8,9 @@
 
 class SimBall:
     def __init__(self):
-        rospy.init_node('ball_sim')
+        rclpy.init(args=sys.argv)
```
Comment (Flova) [inline bitbots_ball_filter/scripts/ball_sim.py]: Missing `rclpy.spin()`, so callback/time/... do not work. See `ball_publisher.py` for correct usage.

**Review by Flova — CHANGES_REQUESTED** [review]

File: bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py (hunk 11 above)
Comment (Flova) [reply bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py:1]: Opinions @val-ba @kindlingAsh ?

Comment (Flova) [reply bitbots_ball_filter/bitbots_ball_filter/ball_publisher.py:1]: Not necessarily relevant for this review, but what do you think?

File: bitbots_ball_filter/scripts/ball_sim.py [hunk 13]
```diff
@@ -49,6 +49,7 @@ def gen_msg(self, x, y):
 
 
 if __name__ == '__main__':
-    SimBall()
+    rclpy.spin(SimBall())
```
Comment (Flova) [inline bitbots_ball_filter/scripts/ball_sim.py]: This should work, but the standard format looks like this:

https://github.com/bit-bots/bitbots_world_model/blob/47e8de43abaab5bdb3f6aa8d85c61078f5e4e494/bitbots_ball_filter/bitbots_ball_filter/ball_filter.py#L258

**Comment by jaagut** [comment]
Closes #27

**Review by Flova — APPROVED** [review]

**Pull request merged**